### ISSUE TITLE
Why no BN in the last few conv layers?

### ISSUE THREAD

**Issue by lionel3** [YudeWang/deeplabv3plus-pytorch#3]
Hi
I have some questions about the Resent-backbone Network.
1. Why no Bn in the last few Conv layers.
2. Why the dilation rates of block-4 of stride16 is (1, 2, 1) rather than (2, 4, 2) as proposed in the paper of deeplabV3. Also, when output stride is 8, the rates should be (4, 8, 4) for block4, am I right?
3. I noticed the dilation setting of your Xception-middle-flow is different from others. Do you have some reference for your setting?
4. How do you normalize your input. I didn't find the normalize part of in you own dataset.

**Comment by YudeWang** [comment]
Thanks for your questions! 

1. In fact, I use SynchronizedBatchNorm2d instead of nn.BatchNorm2d, you can check it in forward( ).
2. There is another deeplabv3 reimplement by MXNet, I read its log file finding that the dilation rate of block-4 of stride8 is (2,4,4), therefore I employ (2,4,2) for stride8 and (1,2,1) for stride16. (https://raw.githubusercontent.com/dmlc/web-data/master/gluoncv/logs/segmentation/deeplab_resnet101_voc.log)
3. I just check the code of official TF deeplabv3+ and the dilation setting of Xception-middle-flow is wrong ( I used to believe the setting of additional layers of Xception should keep the same with resnet). There is no multi-grid dilation rate for xception.
(search the key word 'multi_grid' in https://github.com/tensorflow/models/blob/master/research/deeplab/common.py   )
4. I normalize origin image into [0,1] by /255. You can find this part in datasets.transform.ToTensor().

The official TF code of deeplabv3+ has lots of differences with paper version, I am really glad you can discuss these details with me. Thanks a lot :-)

**Comment by lionel3** [comment]
Thanks for your reply. 
1. I noticed that you use SynchronizedBN instead of BN, I am just confused about the following part.
   
```
 self.shortcut_conv = nn.Sequential(
                nn.Conv2d(indim, cfg.MODEL_SHORTCUT_DIM, cfg.MODEL_SHORTCUT_KERNEL, 1,
                          padding=cfg.MODEL_SHORTCUT_KERNEL // 2),
                nn.ReLU(inplace=True),
            )
            self.cat_conv = nn.Sequential(
                nn.Conv2d(cfg.MODEL_ASPP_OUTDIM + cfg.MODEL_SHORTCUT_DIM, 
    cfg.MODEL_ASPP_OUTDIM, 3, 1, padding=1),
                nn.ReLU(inplace=True),
                nn.Conv2d(cfg.MODEL_ASPP_OUTDIM, cfg.MODEL_ASPP_OUTDIM, 3, 1, padding=1),
                nn.ReLU(inplace=True),
            )
```

It seems that you forgot to add SynBN after the three conv blocks above.

2.  It seems different implements have different settings, I think it depends on the dataset. I followed the setting of this link before [https://github.com/jfzhang95/pytorch-deeplab-xception/blob/master/modeling/backbone/resnet.py](https://github.com/jfzhang95/pytorch-deeplab-xception/blob/master/modeling/backbone/resnet.py)  
3. I checked the official TF codes and also found that the setting of multi-grid dilation rate for xception is [1, 1, 1], it's quite strange.
4. Should we add (image - mean) / std for each channel and make the inputs into [-1, 1].
    
You really have done nice work, happy to discuss the details with you.

**Comment by YudeWang** [comment]
Thanks for your reply!

1. It is true that I forget the BN in decoder (including shortcut and other conv in up-path) at first. However, when I add BN in decoder, the final mIoU drops a little. Therefore I do not use BN in decoder.
2. I will evaluate these settings in future work.
3. May be multi-grid is just designed for ResNet.
4. (image - mean)/std is real normalization, I will test the result of it in next few days.

In fact, this reimplementation is just a part of my work. I wil release others after finishing my paper. Thanks for your suggestions!

**Comment by lionel3** [comment]
Thanks, looking forward to your new work.

**Comment by ImMrMa** [comment]
Wow, it confused me that you didn't let any BN layer to be learned. After seeing this comment, I will try to train it in your cfg files.
Thanks!!
